### DISCUSSION TITLE
NPM Auto Publish

### DISCUSSION THREAD

**Discussion by SamWSoftware** [SamWSoftware/serverless-auto-swagger#25] [Ideas]
How to make sure that we're doing release correctly? 

So far I've always used `np` for publishing. Unfortunately, you can't use this in CICD as it is an interactive CLI tool.

The thing I'm nervous about is getting errors if we publish without bumping the version correctly. 

My ideas are:

1. I give you @bfaulk96 access to the NPM package so you can also deploy manually using np.

2. We use something like https://github.com/marketplace/actions/npm-publish which won't publish if we haven't changed the package number, and has some version-checking functionality.

**Comment by bfaulk96** [comment]
@SamWSoftware for releases, I had a thought that might work well for keeping GH and NPM in line on releases:

1. Merge PRs without version updates
2. When we have a release ready (can be on a per-PR basis or a group), we make an additional update to master with `npm run prepare-release -- --release-as <version>`, which does the following:
  a. Updates the version
  b. Adds a git tag to that commit for the given version
  c. Updates the `CHANGELOG.md` automatically
3. If necessary, adjust the `CHANGELOG.md` file as desired
  a. then run `git commit -v -a --no-edit --amend` (which adds those changes to our versioned commit)
  b. Update tag with `git tag -f <same-tag-name>` (tag name will be something like `v1.3.0`)
5. Push that commit and tag (`git push --follow-tags`.. Might have to also `git push --tags` if tags don't push for some reason)
6. Create a release in GH using that tag (and can copy/paste the `CHANGELOG.md` info into the release)
  a. Doing this will automatically kick off the npm publish action, which will do the work of publishing to NPM for us 😄

What are your thoughts on this?

**Reply by bfaulk96** [reply]
I went ahead and switched the current GH workflow to your suggested (2) idea for now.
It will require `secrets.NPM_TOKEN` to work, similar to the other one.

**Reply by bfaulk96** [reply]
Building off of this, we could potentially have `releases/` branches in the future? And whenever we're ready to release what's in master, we could push to that branch, which could then automatically create the GH release from our changelog and publish to NPM? That would simplify the above flow even more (by removing step 5 essentially)
